### PR TITLE
Add a script to validate jupiter notebooks

### PR DESCRIPTION
**Pull request by abhiram6121**
### Description
 Add Jupyter notebooks to CI #726 

Checklist:
- [x] I have added tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.

### CODE REVIEW

**Comment by abhiram6121** [comment]
I have read the Contributor License Agreement and I hereby accept the Terms.

**Comment by bmhowe23** [comment]
Thanks, @abhiram6121. Please take a look at https://github.com/NVIDIA/cuda-quantum/actions/runs/6721623115?pr=848 to see a couple of things that our CI is pointing out ... namely the missing license preamble and Python formatting. The license preamble can be copied from other Python files, and that prior link provides a diff that can be used to patch your changes (simply formatting updates).

Also, take a look at https://github.com/NVIDIA/cuda-quantum/actions/runs/6722600981/job/18271591978?pr=848. It looks like the new script reported the following:
```
Traceback (most recent call last):
  File "/home/cudaq/notebook_validation.py", line 4, in <module>
    import nbformat
ModuleNotFoundError: No module named 'nbformat'
```

**Comment by abhiram6121** [comment]
> Also, take a look at https://github.com/NVIDIA/cuda-quantum/actions/runs/6722600981/job/18271591978?pr=848. It looks like the new script reported the following:
> 
> ```
> Traceback (most recent call last):
>   File "/home/cudaq/notebook_validation.py", line 4, in <module>
>     import nbformat
> ModuleNotFoundError: No module named 'nbformat'
> ```

It seems `nbconvert` is not installed in the github runner. How do i install it?

**Comment by bettinaheim** [comment]
> > Also, take a look at https://github.com/NVIDIA/cuda-quantum/actions/runs/6722600981/job/18271591978?pr=848. It looks like the new script reported the following:
> > ```
> > Traceback (most recent call last):
> >   File "/home/cudaq/notebook_validation.py", line 4, in <module>
> >     import nbformat
> > ModuleNotFoundError: No module named 'nbformat'
> > ```
> 
> It seems `nbconvert` is not installed in the github runner. How do i install it?

@abhiram6121 In docker_images.yml, after the line you added, you need to add another line
```
docker exec -e TERM=xterm cuda-quantum python3 -m pip install nbconvert
```

**Comment by bettinaheim** [comment]
@abhiram6121 I used your script to validate the notebooks prior to release. Unfortunately, it looks like running the preprocessor doesn't actually execute the notebook code. I know the docs page of nbconvert suggest to do exactly what you did, and states that it will be executed, but that actually seems to be misleading. The docs were not particularly helpful in finding out what function the command line actually invokes, so I would have had to look through the code to see how the API has to be invoked. I instead opted to keep things simpler and made the following updates to your script:

`import glob, os, sys
import subprocess
from shutil import which
if which('jupyter') is None:
    print("Please install jupyter, e.g. with `pip install notebook`.")
    exit(1)


"""
This script runs all the notebooks in the current directory, and catches and reports any raised exceptions.
"""

def execute(notebook_filename):
    notebook_filename_out = notebook_filename.replace(
        '.ipynb', '.nbconvert.ipynb')

    success = True
    try:
        subprocess.run(["jupyter", "nbconvert", "--to", "notebook", "--execute", notebook_filename], check = True)
        os.remove(notebook_filename_out)
    except subprocess.CalledProcessError:
        print('Error executing the notebook "%s".\n\n' % notebook_filename)
        success = False
    return success


if __name__ == '__main__':
    if len(sys.argv) > 1:
        notebook_filenames = sys.argv[1:]
    else:
        # Use glob to find files with the specified extension in the current directory
        notebook_filenames = [
            fn for fn in glob.glob(f"*.ipynb") if not fn.endswith('.nbconvert.ipynb')]

    if notebook_filenames:
        notebooks_failed = []
        for notebook_filename in notebook_filenames:
            if not execute(notebook_filename):
                notebooks_failed.append(notebook_filename)

        if len(notebooks_failed) > 0:
            print("Failed! The following notebook(s) raised errors:\n" +
                  " ".join(notebooks_failed))
        else:
            print("Success! All the notebook(s) executed successfully.")
    else:
        print('Failed! No notebook found in the current directory.')
`

**Comment by abhiram6121** [comment]
@bettinaheim Thanks for the help, have updated accordingly. Please take a look and let me know.

File: docs/notebook_validation.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+def available():
+    available_backends = sys.stdin.readlines()
+    if available_backends:
+        available_backends = [
+            available_backend.strip()
+            for available_backend in available_backends
+        ]
+        return available_backends
+
+def validate(notebook_filename):
```
Comment (bettinaheim) [inline docs/notebook_validation.py:26]: I think you meant update the validate function here to call execute.

Comment (abhiram6121) [reply docs/notebook_validation.py]: yeah, i forgot to call execute function

File: docs/notebook_validation.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        for notebook_filename in notebook_filenames:
+            if (execute(notebook_filename)):
+                notebooks_success.append(notebook_filename)
+            else:
+                notebooks_failed.append(notebook_filename)
+        print_results(notebooks_success, notebooks_failed)
+    else:
+        available_backends = read_available_backends()
+        notebook_filenames = [
+            fn for fn in glob.glob(f"{Path(__file__).parent}/docs/**/*.ipynb",
```
Comment (bettinaheim) [inline docs/notebook_validation.py:84]:
```suggestion
            fn for fn in glob.glob(f"{Path(__file__).parent}/**/*.ipynb",
```

File: docs/notebook_validation.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    if success:
+        print("Success! The following notebook(s) executed successfully:\n" +
+              " ".join(success))
+
+    if failed:
+        print("Failed! The following notebook(s) raised errors:\n" +
+              " ".join(failed))
+
+    if skipped:
+        print("Skipped! The following notebook(s) skipped:\n" +
```
Comment (bettinaheim) [inline docs/notebook_validation.py:63]: This format will make GitHub recognize the warning and show it in the workflow summary.
```suggestion
        print("::warning::Skipped validation for the following notebook(s):\n" +
```

File: docs/notebook_validation.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        return False
+
+
+def print_results(success, failed, skipped=[]):
+    if success:
+        print("Success! The following notebook(s) executed successfully:\n" +
+              " ".join(success))
+
+    if failed:
+        print("Failed! The following notebook(s) raised errors:\n" +
```
Comment (bettinaheim) [inline docs/notebook_validation.py:59]:
```suggestion
        print("::error::The following notebook(s) raised one or more errors:\n" +
```

File: docs/notebook_validation.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+              " ".join(failed))
+
+    if skipped:
+        print("Skipped! The following notebook(s) skipped:\n" +
+              " ".join(skipped))
+
+    if not failed and not skipped:
+        print("Success! All the notebook(s) executed successfully.")
```
Comment (bettinaheim) [inline docs/notebook_validation.py:67]: A non-zero exit code will make the workflow fail. Correspondingly, you want to call exit(1) when there was a notebook that failed.
```suggestion
        print("Success! All the notebook(s) executed successfully.")
    elif failed:
```

File: .github/workflows/docker_images.yml [hunk 6]
```diff
@@ -706,6 +706,8 @@ jobs:
           chown $(id -u):$(id -g) scripts/validate_container.sh
           docker run --rm -dit --name cuda-quantum $cudaq_image
           docker cp scripts/validate_container.sh cuda-quantum:"/home/cudaq/validate_container.sh"
+          docker cp docs/notebook_validation.py cuda-quantum:"/home/cudaq/notebook_validation.py"
+          docker exec -e TERM=xterm cuda-quantum python3 -m pip install glob2 nbconvert nbformat ipykernel
```
Comment (bettinaheim) [inline .github/workflows/docker_images.yml:710]:
```suggestion
          docker exec -e TERM=xterm cuda-quantum python3 -m pip install nbconvert nbformat ipykernel
```

File: docs/sphinx/examples/python/tutorials/noisy_simulations.ipynb [hunk 7]
```diff
@@ -53,7 +53,7 @@
     "kernel = cudaq.make_kernel()\n",
     "qubits = kernel.qalloc(qubit_count)\n",
     "\n",
-    "kernel.x(qubits[0])\n",
+    "kernel.x(qubit[0])\n",
```
Comment (bettinaheim) [inline docs/sphinx/examples/python/tutorials/noisy_simulations.ipynb:56]:
```suggestion
    "kernel.x(qubits[0])\n",
```

File: scripts/validate_container.sh [hunk 8]
```diff
@@ -170,4 +170,6 @@ echo "Total passed: $passed"
 echo "Total failed: $failed"
 echo "Skipped: $skipped"
 echo "============================="
-if [ "$failed" -eq "0" ]; then exit 0; else exit 10; fi
\ No newline at end of file
+
+echo "$available_backends" | python3 notebook_validation.py
```
Comment (bettinaheim) [inline scripts/validate_container.sh:174]:
```suggestion
echo "$available_backends" | "${PYTHONPATH_PREFIX}python3" notebook_validation.py
```

File: .github/workflows/docker_images.yml [hunk 9]
```diff
@@ -703,10 +703,14 @@ jobs:
           fi
 
           # Workaround for dockerd rootless
-          chown $(id -u):$(id -g) scripts/validate_container.sh
           docker run --rm -dit --name cuda-quantum $cudaq_image
```
Comment (bettinaheim) [inline .github/workflows/docker_images.yml:706]:
```suggestion
          chown $(id -u):$(id -g) scripts/validate_container.sh
          docker run --rm -dit --name cuda-quantum $cudaq_image
```

File: .github/workflows/docker_images.yml [hunk 10]
```diff
@@ -703,10 +703,14 @@ jobs:
...
           docker cp scripts/validate_container.sh cuda-quantum:"/home/cudaq/validate_container.sh"
-          docker exec -e TERM=xterm cuda-quantum bash validate_container.sh > /tmp/validation.out
+          docker cp docs/notebook_validation.py cuda-quantum:"/home/cudaq/notebook_validation.py"
+          docker exec -e TERM=xterm cuda-quantum bash -c "sudo apt-get install -y --no-install-recommends wget ca-certificates"
+          docker exec -e TERM=xterm cuda-quantum bash -c \
+            'mkdir -p ~/.miniconda3 && wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-$(uname -m).sh -O ~/.miniconda3/miniconda.sh && bash ~/.miniconda3/miniconda.sh -b -u -p ~/.miniconda3'
+          docker exec -e TERM=xterm cuda-quantum bash -c '~/.miniconda3/bin/conda init && ~/.miniconda3/bin/conda install -y psutil nbconvert nbformat ipykernel'
+          docker exec -e TERM=xterm cuda-quantum bash -c '~/.miniconda3/bin/conda run /home/cudaq/validate_container.sh' > /tmp/validation.out
```
Comment (bettinaheim) [inline .github/workflows/docker_images.yml:713]:
```suggestion
          docker exec -e TERM=xterm cuda-quantum bash -c 'export PYTHONPATH_PREFIX="~/.miniconda3/bin/conda run " && bash validate_container.sh' > /tmp/validation.out
```

**Review by bettinaheim — APPROVED** [review]
At long last, I've updated out images to contain the necessary tools, and everything looks good. [abhiram6121](https://github.com/abhiram6121) Thank you very much for this contribution!

**Pull request merged**